crates/polars-plan/src/plans/aexpr/utils.rs

```diff
@@ -1,121 +1,144 @@
-use bitflags::bitflags;
-
 use super::*;
 
-fn has_series_or_range(ae: &AExpr) -> bool {
-    matches!(
-        ae,
-        AExpr::Literal(LiteralValue::Series(_) | LiteralValue::Range { .. })
-    )
-}
+/// Checks if the top-level expression node is elementwise. If this is the case, then `stack` will
+/// be extended further with any nested expression nodes.
+pub fn is_elementwise(stack: &mut Vec<Node>, ae: &AExpr, expr_arena: &Arena<AExpr>) -> bool {
+    use AExpr::*;
 
-bitflags! {
-        #[derive(Default, Copy, Clone)]
-        struct StreamableFlags: u8 {
-          const ALLOW_CAST_CATEGORICAL = 1;
-        }
-}
+    if !ae.is_elementwise_top_level() {
+        return false;
+    }
 
-#[derive(Copy, Clone)]
-pub struct IsStreamableContext {
-    flags: StreamableFlags,
-    context: Context,
-}
+    match ae {
+        // Literals that aren't being projected are allowed to be non-scalar, so we don't add them
+        // for inspection. (e.g. `is_in(<literal>)`).
+        #[cfg(feature = "is_in")]
+        Function {
+            function: FunctionExpr::Boolean(BooleanFunction::IsIn),
+            input,
+            ..
+        } => (|| {
+            if let Some(rhs) = input.get(1) {
+                assert_eq!(input.len(), 2); // A.is_in(B)
+                let rhs = rhs.node();
 
-impl Default for IsStreamableContext {
-    fn default() -> Self {
-        Self {
-            flags: StreamableFlags::all(),
-            context: Default::default(),
-        }
+                if matches!(expr_arena.get(rhs), AExpr::Literal { .. }) {
+                    stack.push_node(input.first().unwrap().node());
+                    return;
```

Comment:
This is the fix for the `when(A.is_in(B))` issue - we don't traverse into RHS literals of `is_in()`. We had this already in predicate pushdown but it has been moved here.